lucene/core/src/java/org/apache/lucene/search/WANDScorer.java

```diff
@@ -447,6 +461,23 @@ private int doNextCompetitiveCandidate() throws IOException {
       }
     }
 
+      // nocommit Do we still need the following given TwoPhaseIterator.matches already performs the check
+//     minCompetitiveScore satisfied, now checks if the doc has enough required number of matches
+//    while (freq < minShouldMatch) {
+//      if (freq + tailSize >= minShouldMatch) {
+//        // a match on doc is still possible, try to
+//        // advance scorers from the tail
+//        advanceTail();
```

Comment:
I don't think we should advanceTail here, which may be costly. We should just make sure that `freq + tailSize >= minShouldMatch` and otherwise call your `else` block.

Reply:
Makes sense. I've updated the code to include this check in the while loop condition.